collect_app/src/main/java/org/odk/collect/android/instancemanagement/LocalInstancesUseCases.kt

```diff
@@ -0,0 +1,104 @@
+package org.odk.collect.android.instancemanagement
+
+import org.odk.collect.android.utilities.FileUtils
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.forms.instances.InstancesRepository
+import timber.log.Timber
+import java.io.File
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+import java.util.TimeZone
+
+object LocalInstancesUseCases {
+    @JvmOverloads
+    @JvmStatic
+    fun createInstanceFileBasedOnFormPath(
+        formDefinitionPath: String,
+        instancesDir: String,
+        clock: () -> Long = { System.currentTimeMillis() }
+    ): File? {
+        val formFileName = formDefinitionPath
+            .substringAfterLast('/')
+            .substringBeforeLast('.')
+
+        return createInstanceFile(
+            formFileName,
+            instancesDir,
+            clock
+        )
+    }
+
+    fun clone(
+        instanceFile: File?,
+        instancesDir: String,
+        instancesRepository: InstancesRepository,
+        clock: () -> Long = { System.currentTimeMillis() }
+    ): Long? {
+        val sourceInstanceFile = instanceFile ?: return null
+        val targetInstanceFile = copyInstanceDir(sourceInstanceFile, instancesDir, clock) ?: return null
+        val sourceInstance = instancesRepository.getOneByPath(sourceInstanceFile.absolutePath) ?: return null
+
+        return instancesRepository.save(
+            Instance.Builder(sourceInstance)
+                .dbId(null)
+                .status(Instance.STATUS_VALID)
+                .instanceFilePath(targetInstanceFile.absolutePath)
+                .build()
+        ).dbId
+    }
+
+    private fun copyInstanceDir(
+        sourceInstanceFile: File,
+        instancesDir: String,
+        clock: () -> Long = { System.currentTimeMillis() }
+    ): File? {
+        val sourceInstanceDir = sourceInstanceFile.parentFile ?: return null
```

Comment:
I think we should just crash instead of returning `null` all the way through this call chain (all the way up to `InstancesDataService#clone`). There's no (non-exceptional) reason I can think of for any of this not to work, so I'd rather see a crash than a silent failure.

Reply:
I think it would be easier to check for null in `InstancesDataService#clone` and throw an exception there instead of editing this chain here and in `LocalInstancesUseCases#clone`. What do you think?